### ISSUE TITLE
commit 5029c12d7e96cc501721fe4324b03db21bdf3173 builds and *thinks it works* on Apple Silicon M1 but decrypts incorrectly

### ISSUE THREAD

**Issue by jonBoone** [rapier1/chacha20-threaded-interface-broken#1]
After installing openssl 1.1.1i via homebrew, I was able to build the sources like this:

```
❯ export LDFLAGS="-L/Users/jb/opt/homebrew/opt/openssl@1.1/lib"
❯ export CPPFLAGS="-I/Users/jb/opt/homebrew/opt/openssl@1.1/include"
❯ ./configure
...
OpenSSH has been configured with the following options:
                     User binaries: /usr/local/bin
                   System binaries: /usr/local/sbin
               Configuration files: /usr/local/etc
                   Askpass program: /usr/local/libexec/ssh-askpass
                      Manual pages: /usr/local/share/man/manX
                          PID file: /var/run
  Privilege separation chroot path: /var/empty
            sshd default user PATH: /usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin
                    Manpage format: doc
                       PAM support: no
                   OSF SIA support: no
                 KerberosV support: no
                   SELinux support: no
              MD5 password support: no
                   libedit support: no
                   libldns support: no
  Solaris process contract support: no
           Solaris project support: no
         Solaris privilege support: no
       IP address in $DISPLAY hack: no
           Translate v4 in v6 hack: no
                  BSD Auth support: no
              Random number source: OpenSSL internal ONLY
             Privsep sandbox style: darwin
                   PKCS#11 support: yes
                  U2F/FIDO support: yes

              Host: arm-apple-darwin20.1.0
          Compiler: cc
    Compiler flags: -g -O2 -pipe -Wunknown-warning-option -Qunused-arguments -Wall -Wextra -Wpointer-arith -Wuninitialized -Wsign-compare -Wformat-security -Wsizeof-pointer-memaccess -Wno-pointer-sign -Wno-unused-parameter -Wno-unused-result -fno-strict-aliasing -mretpoline -D_FORTIFY_SOURCE=2 -ftrapv -fno-builtin-memset -fstack-protector-strong -fPIE
Preprocessor flags: -I/Users/jb/opt/homebrew/opt/openssl@1.1/include
      Linker flags: -L/Users/jb/opt/homebrew/opt/openssl@1.1/lib -fstack-protector-strong -pie
         Libraries: -lcrypto -lz  -lresolv
         +for sshd:  -lsandbox`
❯  make
...
❯ uname -a
Darwin jbs-MBP-M1.hsd1.pa.comcast.net 20.1.0 Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:10 PDT 2020; root:xnu-7195.50.7~2/RELEASE_ARM64_T8101 arm64
❯ pwd
/Users/jb/src/github.com/rapier1/chacha20-threaded-interface
❯ ./scp -S ./ssh ssh.c localhost:.
1: len = 24, aadlen = 4 seqnr= 3
Exiting chunk loop
Exiting function loop
1: len = 152, aadlen = 4 seqnr= 3
Exiting chunk loop
Exiting function loop
1: len = 24, aadlen = 4 seqnr= 4
Exiting chunk loop
Exiting function loop
1: len = 40, aadlen = 4 seqnr= 4
Exiting chunk loop
Exiting function loop
1: len = 56, aadlen = 4 seqnr= 5
Exiting chunk loop
Exiting function loop
1: len = 600, aadlen = 4 seqnr= 5
Exiting chunk loop
Exiting function loop
1: len = 568, aadlen = 4 seqnr= 6
Exiting chunk loop
Exiting function loop
1: len = 1136, aadlen = 4 seqnr= 6
Exiting chunk loop
Exiting function loop
1: len = 8, aadlen = 4 seqnr= 7
Exiting chunk loop
Exiting function loop
1: len = 32, aadlen = 4 seqnr= 7
Exiting chunk loop
Exiting function loop
1: len = 40, aadlen = 4 seqnr= 8
Exiting chunk loop
Exiting function loop
1: len = 608, aadlen = 4 seqnr= 8
Exiting chunk loop
Exiting function loop
1: len = 128, aadlen = 4 seqnr= 9
Exiting chunk loop
Exiting function loop
1: len = 128, aadlen = 4 seqnr= 10
Exiting chunk loop
Exiting function loop
1: len = 24, aadlen = 4 seqnr= 11
Exiting chunk loop
Exiting function loop
1: len = 32, aadlen = 4 seqnr= 9
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 12
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 13
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 14
Exiting chunk loop
Exiting function loop
1: len = 32, aadlen = 4 seqnr= 10
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 15
Exiting chunk loop
Exiting function loop
ssh.c                                                           0%    0     0.0KB/s   0.0KB/s   --:-- ET1: len = 32784, aadlen = 4 seqnr= 11
1: len is > chunk
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 0, Chunk: 8196 Diff: 32784
3: len-buftr > chunk
bufptr is 0 of 32784 diff 32784
5: leaving chunking
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 8196, Chunk: 8196 Diff: 24588
3: len-buftr > chunk
bufptr is 8196 of 32784 diff 24588
5: leaving chunking
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 16392, Chunk: 8196 Diff: 16392
3: len-buftr > chunk
bufptr is 16392 of 32784 diff 16392
5: leaving chunking
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 24588, Chunk: 8196 Diff: 8196
3: len-buftr > chunk
bufptr is 24588 of 32784 diff 8196
bfptr and len match
5: leaving chunking
6: building structs
6: building structs
6: building structs
6: building structs
0 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
0 of 4 6165491712 JOIN
Exiting join loop
total: 4 joined 1
1 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
1 of 4 6165491712 JOIN
Exiting join loop
total: 4 joined 2
2 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
2 of 4 6165491712 JOIN
Exiting join loop
total: 4 joined 3
3 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
3 of 4 6165491712 JOIN
Exiting join loop
total: 4 joined 4
Exiting chunk loop
Exiting function loop
1: len = 32784, aadlen = 4 seqnr= 12
1: len is > chunk
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 0, Chunk: 8196 Diff: 32784
3: len-buftr > chunk
bufptr is 0 of 32784 diff 32784
5: leaving chunking
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 8196, Chunk: 8196 Diff: 24588
3: len-buftr > chunk
bufptr is 8196 of 32784 diff 24588
5: leaving chunking
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 16392, Chunk: 8196 Diff: 16392
3: len-buftr > chunk
bufptr is 16392 of 32784 diff 16392
5: leaving chunking
2: bufptr < len
aad: 4 Len: 32784, Buffptr: 24588, Chunk: 8196 Diff: 8196
3: len-buftr > chunk
bufptr is 24588 of 32784 diff 8196
bfptr and len match
5: leaving chunking
6: building structs
6: building structs
6: building structs
6: building structs
0 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
0 of 4 6165491712 JOIN
Exiting join loop
total: 8 joined 5
1 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
1 of 4 6165491712 JOIN
Exiting join loop
total: 8 joined 6
2 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
2 of 4 6165491712 JOIN
Exiting join loop
total: 8 joined 7
3 of 4 6165491712 MADE
Made thread!
Leaving 6165491712
3 of 4 6165491712 JOIN
Exiting join loop
total: 8 joined 8
Exiting chunk loop
Exiting function loop
1: len = 7544, aadlen = 4 seqnr= 13
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 16
Exiting chunk loop
Exiting function loop
ssh.c                                                         100%   71KB  30.1MB/s  71.4KB/s   00:00
1: len = 16, aadlen = 4 seqnr= 14
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 17
Exiting chunk loop
Exiting function loop
1: len = 32, aadlen = 4 seqnr= 18
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 19
Exiting chunk loop
Exiting function loop
1: len = 16, aadlen = 4 seqnr= 15
Exiting chunk loop
Exiting function loop
1: len = 40, aadlen = 4 seqnr= 16
Exiting chunk loop
Exiting function loop
❯  diff ./ssh.c ~/ssh.c
279c279
< 	    "d", homedir,
---
> 	    "d", hir,
557,558c557
<  notfound:
< 	if (!options.canonicalize_fallback_local)
---
>  notfoundf (!options.canonicalize_fallback_local)
859,860c858
< 			free(p);
< 			break;
---
> 			free(			break;
1427c1425
< 		free(options.identity_agent);
---
> 		free(options.iity_agent);
1683c1681
< 	for (i = 0; i < options.num_certificate_files; i++) {
---
> 	for (i = 0; options.num_certificate_files; i++) {
1945c1943
< 		if ((options.remote_forwards[i].handle =
---
> 		if ((options.remote_forwards[i].han=
```

**Comment by jonBoone** [comment]
I think I have discovered the root cause of the issue.  Each thread struct is pointing to the same shared u_char* dest location.  

Mallocing a thread local dest and then copying the results from the threads into the shared dest should fix it.

**Comment by jonBoone** [comment]
After further investigation, I think the original approach was better -- Mac OS does not implement pthread synchronization barriers, so unless we were to do so ourselves, a simple mutex on the shared dest buffer is probably best.

**Comment by rapier1** [comment]
Any idea where to do that? I've placed locks around the cipher process in 
the function called by the thread. Same issue. I've also tried locks in 
various other places (such as when I'm populating the struct). I'll commit 
another version for your review.

**Comment by jonBoone** [comment]
> Any idea where to do that? I've placed locks around the cipher process in the function called by the thread. Same issue. I've also tried locks in various other places (such as when I'm populating the struct). I'll commit another version for your review.

Claiming the mutex inside chachapoly_thread_work() immediately before and immediately after the call to EVP_CIPHER() should be the correct place to prevent the threads from stepping on each other.

From the debugging I've done (mostly refining the fprintf() statements to make things clearer), it seems that the issue may be that the main thread ends before the worker threads are able to start and perform their tasks.

There are a few ways I can think of to try to tackle this issue:
1. Have the main thread wait() until all of the worker threads have completed their tasks, then perform the final tag calculation (if encrypting) and exit the chachapoly_crypt() function.  The problem with this is that we'd have to add small sleep()s to keep from chewing up CPU waiting on the worker threads to finish.
2. Have the main thread act as the worker thread on the 0th chunk (just like it does in the case where the src buffer is smaller than the size of a chunk), while spinning up additional workers to do any additional chunks.  I'm not sure if this will fix the problem entirely, but it will at least give the other worker threads an opportunity to start running and do some computations before the main thread moves on.
3.  Some combination of the two approaches.

I'm having dinner -- and then I have to rebuild my Linux box (long story, longer process) -- but I'll try to take a glance at the new version you have checked-in no later than 09:00 EST tomorrow morning.

--jon

**Comment by jonBoone** [comment]
Issued a PR with the enhanced debug messages -- feel free to accept/reject at your discretion.

**Comment by jonBoone** [comment]
With respect to approach 1., I think that using a pthread conditional variable will do the trick nicely (example at: https://randu.org/tutorials/threads/).  In the current implementation, I'd have the chachaoply_thread_work() increment the conditional variable by 1 after it's successfully processed the local_thread.src data into local_thread.dest.

In the main thread (which would not be a worker itself), have it wait on the conditional variable until it is equal to the number of threads started.

**Comment by rapier1** [comment]
So let me see if I understand this.

threadfunc {
   mutex lock
     dowork ()
     counter++
   mutex unlock
}

main {

for (i .. maxthreads) {
   pthread_create (threadfunc)
   pthread_join
}
pthread_condwait (counter == maxthreads)

}

Is that about right?

**Comment by jonBoone** [comment]
I think you've understood, but to be certain, let me rephrase it:

insert counter++ line [here](https://github.com/rapier1/chacha20-threaded-interface/blob/7e02a949efcfc9f094107e244e3ae2eb10b00501/cipher-chachapoly-libcrypto.c#L104)

insert pthread_condwait(counter==maxthreads) [here](https://github.com/rapier1/chacha20-threaded-interface/blob/7e02a949efcfc9f094107e244e3ae2eb10b00501/cipher-chachapoly-libcrypto.c#L394)

Then we can work on cleaning up the thread structure creation, etc.  It really should all be one loop... :-)

**Comment by rapier1** [comment]
Yup, did that. Still hitting some problems but I think it's a new set of 
problems now. BTW: I did have to change the counter=maxthreads because 
there will be times when it will only spin off 1 to 3 threads. I'll 
commit what I've done.

Oh, it was all one loop at one point but I started breaking it down into 
different loops when I couldn't figure shit out :)

Cuz increased complexity is the right way to isolate and solve problems. :)

**Comment by jonBoone** [comment]
Cool!

It actually should be that the pthread_conditional == number of threads, so you’re right to change that!

Also, try increasing chunk=8196 (that probably works slightly better).

**Comment by jonBoone** [comment]
One more thing before you commit your changes -- there's a loop where we are copying data from the thread[i].dest to the *dest that was passed into the function -- please delete that -- it was from another experiment I was doing and shouldn't be in the code you have.

**Comment by jonBoone** [comment]
Issued a new PR -- with these changes, you should see that an scp'd file is very similar, though not exactly the same, as the original.

**Comment by rapier1** [comment]
@jonBoone Hey, I saw the PR for the confusion with the dest copy. I fixed that prior to this as it looked like you just forgot about it. Is that the pull request you were talking about or is there another one? Sorry for the delay in getting back to you. This day has been a shitshow. I'll be doing some work tonight and getting back in the groove tomorrow.

**Comment by rapier1** [comment]
I pushed a new commit. It's not working as I thought it would. It getting hung up somewhere. I haven't done anything with gdb yet as I have to make dinner.

**Comment by jonBoone** [comment]
For sanity's sake, I have a branch in my repo called `restore-original`, in which I replaced `cipher-chachapoly-libcrypto.c` with `cipher-chachapoly-libcrypto.c.good`.  Is that the proper way to restore the src back to its initial state (prior to any changes you started making)? 

Also, I've been reading [RFC 8439 - ChaCha20 and Poly1305 for IETF Protocols](https://tools.ietf.org/html/rfc8439) and, unless I'm misunderstanding, the changes we are attempting to implement will *not* work without a similar implementation on the other side.  The short story is, each time we call EVP_Cipher() we are passing in what the cipher implementation thinks is the entire packet, so the resulting cryptotexts can't be combined together, but must be sent separately, which will kill your throughput on the wire.  Please review the RFC and let me know if you disagree.

It seems like our best bet at this time is to take the threading work we've been trying to add to `cipher-chacha20poly-libcrypto.c` and work it into `cipher-chachapoly.c`.

**Comment by rapier1** [comment]
On 12/17/20 8:10 AM, Jon Boone wrote:
> For sanity's sake, I have a branch in my repo called |restore-original|, 
> in which I replaced |cipher-chachapoly-libcrypto.c| with 
> |cipher-chachapoly-libcrypto.c.good|. Is that the proper way to restore 
> the src back to its initial state (prior to any changes you started making)?

Honestly, I don't know I'll take a look.

> Also, I've been reading RFC 8439 - ChaCha20 and Poly1305 for IETF 
> Protocols <https://tools.ietf.org/html/rfc8439> and, unless I'm 
> misunderstanding, the changes we are attempting to implement will /not/ 
> work without a similar implementation on the other side. The short story 
> is, each time we call EVP_Cipher() we are passing in what the cipher 
> implementation thinks is the entire packet, so the resulting cryptotexts 
> can't be combined together, but must be sent separately, which will kill 
> your throughput on the wire. Please review the RFC and let me know if 
> you disagree.

I believe that would be true if each time we sent data to the EVP_Cipher 
we reinitialized the cipher context. However, we only do that once so 
the nonce should be consistent across multiple blocks of data (assuming 
that assuming that the ctx is actually, ya know, being copied 
correctly). As a counter cipher it should be possible, as long as the 
nonce remains consistent, generate any arbitrary block for that counter 
as part of a stream. The question is if the openssl implementation 
actually accepts arbitrary block counters. I've been assuming that they do.

Can you point to where in the RFC you think I should be looking? To be 
honest I haven't even thought about reading this RFC as I've mostly just 
been reading Bernstein's papers.

One of the things I've been working on today is trying to assure that 
the way I'm memcpying src to the thread struct is correct. I don't think 
it is. I believe I'm, in places, screwing up the boundaries causing the 
bytestream to be off by x bytes (which is why garbage appears in some 
transfers as the xor would be off). In other places I think I'm leaving 
gaps which is causing the 4 bytes of null characters to get inserted at 
what seem to be chunk boundaries.

Even if this doesn't work with the libcrypto version getting the buffer 
math correct is important so...

Honestly, what I'd like to do is what Ben Bennet (a colleague) 
implemented for the AES-CTR cipher (cipher-ctr-mt.c) which created a 
precomputed keysteam queue that we could just xor the data stream 
against. That might be the right way to do things as you can set up an 
arbitrary number of threads (which is good when dealing with processors 
with 64 cores) but I wanted to do this as some level of proof that 
threading chacha20 would provide some level of benefit in terms of 
throughput.

> It seems like our best bet at this time is to take the threading work 
> we've been trying to add to |cipher-chacha20poly-libcrypto.c| and work 
> it into |cipher-chachapoly.c|.

I've been looking at that but it's not used if openssl is enabled during 
compilation so I decided to start with this one. It's not that we can't 
use that one it's just more fiddling with other bits of code in other 
files.

By way, I *really* appreciate all of this.

**Comment by rapier1** [comment]
Something I just thought of -

The reason why it may work with sequential threads as a opposed to 
concurrent is that the OpenSSL version of chacha20 may not be thread 
safe. I'm looking into that now and think there may be a solution using 
what are called cryptolocks. 
https://www.openssl.org/blog/blog/2017/02/21/threads/

Again, this might be a dead end but can't hurt to try it.

**Comment by rapier1** [comment]
And this 
https://www.openssl.org/docs/man1.1.1/man3/CRYPTO_THREAD_lock_new.html

**Comment by jonBoone** [comment]
_> I believe that would be true if each time we sent data to the EVP_Cipher
we reinitialized the cipher context. However, we only do that once so
the nonce should be consistent across multiple blocks of data (assuming
that assuming that the ctx is actually, ya know, being copied
correctly). As a counter cipher it should be possible, as long as the
nonce remains consistent, generate any arbitrary block for that counter
as part of a stream. The question is if the openssl implementation
actually accepts arbitrary block counters. I've been assuming that they do.

Can you point to where in the RFC you think I should be looking? To be
honest I haven't even thought about reading this RFC as I've mostly just
been reading Bernstein's papers._

My confusion (and it may just be in my head) is around 
[the actual invocation of EVP_Cipher()](https://github.com/rapier1/chacha20-threaded-interface/blob/a2d459b0ea417ec8b29e951c4f28e90ee2615b2d/cipher-chachapoly-libcrypto.c#L106).  Notice that when we are passing in the src and dest locations, we're shifting them, not only by the current buffer position, but also by the aadlen.  My understanding, per the RFC, is that aad only comes into play after all of the plaintext has been encrypted and as part of the generation of the tag.  See step 3 of [RFC 8439, section 2.8](https://tools.ietf.org/html/rfc8439#section-2.8).

So, it seems to me, that referencing space for the aad makes sense only in one of the two following scenarios:

1. The chunk is an even multiple of 16 bytes and aadlen = 0
2. The chunk is < an even multiple of 16 bytes and aadlen + sizeof(chunk) is an even multiple of 16.

But we're always setting the same aadlen value for each invocation, which doesn't seem to be correct to me.

**Comment by rapier1** [comment]
I think I fixed that in the latest set of work. I haven't committed it 
yet though. I'll take a look on the 26th or 27th and see where it stands.

**Comment by jonBoone** [comment]
I've compiled the code in the `cryptolocks` branch.  The good news is that it works some of the time when I scp the `ssh.c` file to $HOME/ssh.c via localhost.  I ran it manually about 6 times and it worked 5 of the 6 times.

The bad news is that the 1 time it didn't work, only about 1/2 of the `ssh.c` file was successfully transferred and the process hung until i canceled it with `^-c`.  Clearly, there's still a timing issue.
